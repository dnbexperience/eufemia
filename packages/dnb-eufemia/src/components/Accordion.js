/**
 * ATTENTION: This file is auto generated by using "prepareTemplates".
 * Do not change the content!
 *
 */

/**
 * Library Index accordion to autogenerate all the components and extensions
 * Used by "prepareAccordions"
 */

import Accordion from './accordion/Accordion'
export * from './accordion/Accordion'
export default Accordion
