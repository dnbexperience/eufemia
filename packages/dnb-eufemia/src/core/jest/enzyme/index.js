/**
 * Default Component Test Setup
 *
 */

// enzyme setup
const enzyme = require('enzyme')

// TODO: This may not be needed in future
const Adapter = require('enzyme-adapter-react-16')
enzyme.configure({ adapter: new Adapter() })

module.exports = enzyme
