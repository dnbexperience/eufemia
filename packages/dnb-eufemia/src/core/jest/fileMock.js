export default 'static file content'
