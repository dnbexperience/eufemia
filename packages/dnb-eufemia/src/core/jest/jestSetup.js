/**
 * Jest Setup
 *
 */

import { axe, toHaveNoViolations } from 'jest-axe'
import fakeProps from 'react-fake-props'
import { shallow, mount, render } from './enzyme'
import ReactDOMServer from 'react-dom/server'
import fs from 'fs-extra'
import onceImporter from 'node-sass-once-importer'
import path from 'path'
import sass from 'node-sass'
import { toBeType } from 'jest-tobetype'
import toJson from 'enzyme-to-json'

export {
  fakeProps, // we have also our own replacement function called "fakeAllProps"
  shallow,
  mount,
  render,
  toJson,
  axe,
  toHaveNoViolations
}

global.IS_TEST = true
if (typeof window !== 'undefined') {
  window.IS_TEST = true
}

expect.extend({ toBeType })
expect.extend(toHaveNoViolations)

export const loadScss = (file, options = {}) => {
  try {
    const sassResult = sass.renderSync({
      file,
      includePaths: [path.resolve(__dirname, '../../style/core/')],
      importer: [onceImporter()],
      ...options
    })
    return String(sassResult.css)
  } catch (e) {
    console.error('loadScss error:', e)
    return e
  }
}

export const mockGetSelection = () => {
  let memory
  Object.defineProperty(window.navigator, 'clipboard', {
    configurable: true,
    value: {
      writeText: jest.fn().mockImplementation((v) => {
        memory = v
        return Promise.resolve(v)
      }),
      readText: jest.fn().mockImplementation(() => Promise.resolve(memory))
    }
  })

  const mockRange = new (class Range {
    constructor() {
      this.startContainer = {
        parentNode: document.createElement('div')
      }
    }
    getElement() {
      return this.node
    }
    insertNode(elem) {
      this.node = document.createElement('div')
      this.node.appendChild(elem)
      return this
    }
    cloneRange() {
      return this
    }
  })()

  let ranges = [mockRange]

  const mockValue = '1234.56'
  let value = mockValue
  let rangeCount = 9

  class RangeObj {
    rangeCount = rangeCount
    toString = () => value
    addRange(range = mockRange) {
      value = mockValue
      ranges.push(range)
      rangeCount = ranges.length
    }
    getRangeAt(index) {
      return ranges[index]
    }
    removeAllRanges() {
      value = ''
      ranges = []
      rangeCount = ranges.length
    }
  }

  Object.defineProperty(window, 'getSelection', {
    configurable: true,
    value: () => {
      return new RangeObj()
    }
  })
}

export const loadImage = async (imagePath) =>
  await fs.readFile(path.resolve(imagePath))

export const toHtml = (Component) =>
  ReactDOMServer.renderToStaticMarkup(Component)

export const axeComponent = async (...components) => {
  const html = components
    .filter((Component) =>
      // enzyme names the mounted wrapper: ReactWrapper
      /react/i.test(String(Component.constructor))
    )
    .map((Component) => toHtml(Component))
    .join('\n')

  return await axe(
    `<main>${html}</main>`,
    typeof components[1] === 'object' ? components[1] : null
  )
}

export function attachToBody() {
  let container = document.createElement('div')
  document.body.append(container)
  return container
}

function jestSetup() {}

export default jestSetup
