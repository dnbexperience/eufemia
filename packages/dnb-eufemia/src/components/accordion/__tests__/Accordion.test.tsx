/**
 * Component Test
 *
 */

import React from 'react'
import { axeComponent, loadScss } from '../../../core/jest/jestSetup'
import Accordion, { AccordionProps } from '../Accordion'
import {
  add_medium as AddIcon,
  subtract_medium as SubtractIcon,
} from '../../../icons'
import { render, fireEvent, act } from '@testing-library/react'
import MatchMediaMock from 'jest-matchmedia-mock'
import userEvent from '@testing-library/user-event'

new MatchMediaMock()

const props: AccordionProps = {
  no_animation: true,
  title: 'title',
}

describe('Accordion component', () => {
  it('has correct state after "click" trigger', () => {
    const { rerender } = render(<Accordion {...props} />)

    // default expanded value has to be false
    expect(
      document
        .querySelector('.dnb-accordion__header')
        .getAttribute('aria-expanded')
    ).toBe('false')
    fireEvent.click(document.querySelector('.dnb-accordion__header')) // we could send inn the event data structure like this: , { target: { expanded: true } }
    expect(
      document
        .querySelector('.dnb-accordion__header')
        .getAttribute('aria-expanded')
    ).toBe('true')
    rerender(<Accordion {...props} expanded={false} />)
    expect(
      document
        .querySelector('.dnb-accordion__header')
        .getAttribute('aria-expanded')
    ).toBe('false')
  })

  it('has "on_change" event which will trigger on click', () => {
    const my_event = jest.fn()
    const myEvent = jest.fn()
    render(
      <Accordion
        {...props}
        on_change={my_event}
        onChange={myEvent}
        expanded={false}
      />
    )

    // first click
    fireEvent.click(document.querySelector('.dnb-accordion__header'))
    expect(my_event).toHaveBeenCalled()
    expect(my_event.mock.calls[0][0].expanded).toBe(true)
    expect(myEvent.mock.calls.length).toBe(1)
    expect(myEvent.mock.calls[0][0]).toHaveProperty('expanded')
    expect(myEvent.mock.calls[0][0].expanded).toBe(true)

    // second click
    fireEvent.click(document.querySelector('.dnb-accordion__header'))
    expect(my_event.mock.calls[1][0].expanded).toBe(false)
  })

  it('uses a p element when string content is given', () => {
    const { rerender } = render(
      <Accordion {...props} expanded>
        string content
      </Accordion>
    )

    expect(document.querySelector('.dnb-p').textContent).toBe(
      'string content'
    )

    rerender(
      <Accordion {...props} expanded>
        <span className="no-string">no string content</span>
      </Accordion>
    )

    expect(document.querySelector('.dnb-p')).toBeFalsy()
    expect(document.querySelector('.no-string').textContent).toBe(
      'no string content'
    )
  })

  it('has a disabled attribute, once we set disabled to true', () => {
    const { rerender } = render(<Accordion {...props} />)

    rerender(<Accordion {...props} disabled={true} />)
    expect(
      document.querySelector('.dnb-accordion__header')
    ).toHaveAttribute('disabled')
  })

  it('has correct classes when no_animation', () => {
    render(<Accordion no_animation />)

    expect(
      Array.from(
        document.querySelector('.dnb-accordion__header').classList
      )
    ).toEqual(
      expect.arrayContaining(['dnb-accordion__header--no-animation'])
    )
  })

  it('supports an icon for expanded state', () => {
    render(
      <Accordion
        {...props}
        icon={{
          closed: AddIcon,
          expanded: SubtractIcon,
        }}
      />
    )

    const { container: c1 } = render(<AddIcon />)
    const { container: c2 } = render(<SubtractIcon />)

    expect(document.querySelector('.dnb-accordion svg').outerHTML).toBe(
      c1.innerHTML
    )

    fireEvent.click(document.querySelector('.dnb-accordion__header'))

    expect(document.querySelector('.dnb-accordion svg').outerHTML).toBe(
      c2.innerHTML
    )
  })

  it('supports default outlined variant', () => {
    render(<Accordion />)
    expect(document.querySelector('.dnb-accordion').classList).toContain(
      'dnb-accordion__variant--outlined'
    )
  })

  it('supports plain variant', () => {
    render(<Accordion {...props} variant="plain" />)
    expect(document.querySelector('.dnb-accordion').classList).toContain(
      'dnb-accordion__variant--plain'
    )
  })

  it('should have hidden content when "prerender" is true but closed', () => {
    render(<Accordion {...props} prerender={true} />)

    expect(
      Array.from(
        document.querySelector('.dnb-accordion__content').classList
      )
    ).toEqual(expect.arrayContaining(['dnb-height-animation--hidden']))

    fireEvent.click(document.querySelector('.dnb-accordion__header'))

    expect(
      Array.from(
        document.querySelector('.dnb-accordion__content').classList
      )
    ).toEqual(expect.arrayContaining(['dnb-height-animation--is-in-dom']))
  })

  it('should validate with ARIA rules', async () => {
    const Comp = render(<Accordion {...props} />)

    expect(await axeComponent(Comp)).toHaveNoViolations()
  })
})

describe('Accordion store API', () => {
  it('will save and read the states for a single accordion', () => {
    const inst = Accordion.Store('accordion-id')

    inst.saveState(true)
    expect(inst.getState()).toBe(true)
    expect(inst.getData()).toMatchObject({ expanded: true })

    inst.saveState(false)
    expect(inst.getState()).toBe(false)
    expect(inst.getData()).toMatchObject({ expanded: false })

    inst.flush()
    expect(inst.getState()).toBe(null)
    expect(inst.getData()).toBe(null)
  })

  it('will save and read the states for an accordion group', () => {
    const inst = Accordion.Group.Store('group-id')

    inst.saveState(true, 'remembered-state-2')
    expect(inst.getState('remembered-state-2')).toBe(true)
    expect(inst.getData()).toMatchObject({ id: 'remembered-state-2' })

    inst.saveState(false, 'remembered-state-2', { force: true })
    expect(inst.getState('remembered-state-2')).toBe(false)
    expect(inst.getData()).toMatchObject({ id: null })

    inst.flush('remembered-state-2')
    expect(inst.getState('remembered-state-2')).toBe(null)
    expect(inst.getData()).toBe(null)
  })
})

describe('Accordion group component', () => {
  it('has to inherit expanded from group', () => {
    render(
      <Accordion.Group expanded id="group">
        <Accordion id="accordion-1" title="Accordion 1">
          Accordion 1
        </Accordion>
        <Accordion id="accordion-2" title="Accordion 2" expanded={false}>
          Accordion 2
        </Accordion>
      </Accordion.Group>
    )

    expect(
      document.querySelector('#accordion-1 .dnb-accordion__content')
    ).toBeInTheDocument()
    expect(
      document.querySelector('#accordion-2 .dnb-accordion__content')
    ).toBeFalsy()
  })

  it('has "on_change" event which will trigger on a button click', () => {
    const my_event = jest.fn()
    const myEvent = jest.fn()
    render(
      <Accordion.Group
        id="group"
        on_change={my_event}
        onChange={myEvent}
        value="second"
        data-prop="group-value"
      >
        <Accordion
          id="accordion-1"
          value="first"
          data-prop="value-1"
          attributes={{ 'data-attr': 'value' }}
        />
        <Accordion
          id="accordion-2"
          value="second"
          data-prop="value-2"
          attributes={{ 'data-attr': 'value' }}
        />
      </Accordion.Group>
    )

    fireEvent.click(
      document.querySelector('#accordion-1 .dnb-accordion__header')
    )
    expect(my_event).toHaveBeenCalled()
    expect(my_event.mock.calls[0][0].id).toBe('accordion-1')
    expect(my_event.mock.calls[0][0].expanded).toBe(true)
    expect(myEvent.mock.calls.length).toBe(1)

    fireEvent.click(
      document.querySelector('#accordion-2 .dnb-accordion__header')
    )
    expect(my_event.mock.calls[1][0].id).toBe('accordion-2')
    expect(my_event.mock.calls[1][0].expanded).toBe(true)

    fireEvent.click(
      document.querySelector('#accordion-1 .dnb-accordion__header')
    )
    expect(my_event.mock.calls[2][0].expanded).toBe(true)
  })

  it('should close all accordions inside a group with collapseAllHandleRef', () => {
    const collapseAll = React.createRef<() => void>()

    render(
      <Accordion.Group
        expanded
        allow_close_all
        collapseAllHandleRef={collapseAll}
      >
        <Accordion>
          <Accordion.Header>Accordion title 1</Accordion.Header>
          <Accordion.Content>
            Sociis sapien sociosqu vel sollicitudin accumsan laoreet
            gravida himenaeos nostra mollis volutpat bibendum convallis cum
            condimentum dictumst blandit rutrum vehicula
          </Accordion.Content>
        </Accordion>
        <Accordion>
          <Accordion.Header>Accordion title 2</Accordion.Header>
          <Accordion.Content>
            Nec sit mattis natoque interdum sagittis cubilia nibh nullam
            etiam
          </Accordion.Content>
        </Accordion>
        <Accordion>
          <Accordion.Header>Accordion title 2</Accordion.Header>
          <Accordion.Content>
            Nec sit mattis natoque interdum sagittis cubilia nibh nullam
            etiam
          </Accordion.Content>
        </Accordion>
      </Accordion.Group>
    )

    const [first, second, third] = Array.from(
      document.querySelectorAll('.dnb-accordion')
    )

    expect(first).toHaveClass('dnb-accordion--expanded')
    expect(second).toHaveClass('dnb-accordion--expanded')
    expect(third).toHaveClass('dnb-accordion--expanded')

    act(collapseAll.current)

    expect(first).not.toHaveClass('dnb-accordion--expanded')
    expect(second).not.toHaveClass('dnb-accordion--expanded')
    expect(third).not.toHaveClass('dnb-accordion--expanded')
  })
})

describe('Accordion container component', () => {
  type DidRenderProps = {
    id: string
  }
  class DidRender extends React.PureComponent<DidRenderProps> {
    state = { mounted: false }
    componentDidMount() {
      this.setState({ mounted: true })
    }
    render() {
      return <div id={this.props.id}>{String(this.state.mounted)}</div>
    }
  }

  const Increment = () => {
    const [count, setCount] = React.useState(1)
    return (
      <button id="increment" onClick={() => setCount((s) => s + 1)}>
        {count}
      </button>
    )
  }

  const Container = (props) => {
    return (
      <>
        <Increment />
        <Accordion.Group
          id="container"
          single_container
          prevent_rerender
          remember_state
          no_animation
          {...props}
        >
          <Accordion id="accordion-1" title="Accordion 1">
            Accordion 1
            <DidRender id="mounted-1" />
          </Accordion>
          <Accordion id="accordion-2" title="Accordion 2" expanded={true}>
            Accordion 2
            <DidRender id="mounted-2" />
          </Accordion>
          <Accordion id="accordion-3" title="Accordion 3">
            Accordion 3
            <DidRender id="mounted-3" />
          </Accordion>
        </Accordion.Group>
      </>
    )
  }

  it('has only to render the expanded accordion content', () => {
    render(<Container />)

    expect(document.querySelector('button#increment').textContent).toBe(
      '1'
    )

    fireEvent.click(document.querySelector('button#increment'))
    expect(document.querySelector('button#increment').textContent).toBe(
      '2'
    )

    expect(document.querySelector('div#mounted-1')).toBeFalsy()
    expect(document.querySelector('div#mounted-2').textContent).toBe(
      'true'
    )
    expect(document.querySelector('div#mounted-3')).toBeFalsy()
    expect(
      document
        .querySelector('#accordion-2 .dnb-accordion__header')
        .getAttribute('aria-expanded')
    ).toBe('true')

    fireEvent.click(
      document.querySelector('#accordion-1 .dnb-accordion__header')
    )

    fireEvent.click(document.querySelector('button#increment'))
    expect(document.querySelector('button#increment').textContent).toBe(
      '3'
    )

    expect(document.querySelector('div#mounted-1').textContent).toBe(
      'true'
    )
    expect(document.querySelector('div#mounted-2').textContent).toBe(
      'true'
    )
    expect(document.querySelector('div#mounted-3')).toBeFalsy()

    fireEvent.click(
      document.querySelector('#accordion-2 .dnb-accordion__header')
    )

    fireEvent.click(document.querySelector('button#increment'))
    expect(document.querySelector('button#increment').textContent).toBe(
      '4'
    )

    expect(document.querySelector('div#mounted-3')).toBeFalsy()

    fireEvent.click(
      document.querySelector('#accordion-3 .dnb-accordion__header')
    )

    fireEvent.click(document.querySelector('button#increment'))
    expect(document.querySelector('button#increment').textContent).toBe(
      '5'
    )

    expect(document.querySelector('div#mounted-3').textContent).toBe(
      'true'
    )
  })

  it('will set minHeight', async () => {
    const contentRef = React.createRef<HTMLElement>()

    render(<Container contentRef={contentRef} />)

    const contentElem = contentRef.current

    jest.spyOn(contentElem, 'offsetHeight', 'get').mockReturnValue(48)
    jest.spyOn(contentElem, 'offsetTop', 'get').mockReturnValue(48)

    fireEvent.click(
      document.querySelector('#accordion-1 .dnb-accordion__header')
    )

    expect(
      document
        .querySelector('.dnb-accordion-group--single-container')
        .getAttribute('style')
    ).toBe('transition-duration: 1ms; min-height: 6rem;')
  })

  it('should allow all accordions inside a group to be expanded at the same time', async () => {
    render(
      <Accordion.Group expandBehaviour="multiple">
        <Accordion>
          <Accordion.Header>Accordion title 1</Accordion.Header>
          <Accordion.Content>
            Sociis sapien sociosqu vel sollicitudin accumsan laoreet
            gravida himenaeos nostra mollis volutpat bibendum convallis cum
            condimentum dictumst blandit rutrum vehicula
          </Accordion.Content>
        </Accordion>
        <Accordion>
          <Accordion.Header>Accordion title 2</Accordion.Header>
          <Accordion.Content>
            Nec sit mattis natoque interdum sagittis cubilia nibh nullam
            etiam
          </Accordion.Content>
        </Accordion>
        <Accordion>
          <Accordion.Header>Accordion title 3</Accordion.Header>
          <Accordion.Content>
            Nec sit mattis natoque interdum sagittis cubilia nibh nullam
            etiam
          </Accordion.Content>
        </Accordion>
      </Accordion.Group>
    )

    const [first, second, third] = Array.from(
      document.querySelectorAll('.dnb-accordion__header')
    )

    expect(first).toHaveAttribute('aria-expanded', 'false')
    expect(second).toHaveAttribute('aria-expanded', 'false')
    expect(third).toHaveAttribute('aria-expanded', 'false')

    await userEvent.click(first)
    expect(first).toHaveAttribute('aria-expanded', 'true')
    expect(second).toHaveAttribute('aria-expanded', 'false')
    expect(third).toHaveAttribute('aria-expanded', 'false')

    await userEvent.click(second)
    expect(first).toHaveAttribute('aria-expanded', 'true')
    expect(second).toHaveAttribute('aria-expanded', 'true')
    expect(third).toHaveAttribute('aria-expanded', 'false')

    await userEvent.click(third)
    expect(first).toHaveAttribute('aria-expanded', 'true')
    expect(second).toHaveAttribute('aria-expanded', 'true')
    expect(third).toHaveAttribute('aria-expanded', 'true')
  })
})

describe('Accordion scss', () => {
  it('has to match style dependencies css', () => {
    const css = loadScss(require.resolve('../style/deps.scss'))
    expect(css).toMatchSnapshot()
  })

  it('have to match default theme snapshot', () => {
    const css = loadScss(
      require.resolve('../style/themes/dnb-accordion-theme-ui.scss')
    )
    expect(css).toMatchSnapshot()
  })
})
