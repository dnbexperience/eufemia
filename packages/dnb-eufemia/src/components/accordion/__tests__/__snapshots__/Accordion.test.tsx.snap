// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordion scss has to match style dependencies css 1`] = `
"/*
* Used for snapshot testing
*
*/
/*
 * Utilities
 */
/*
* Used for snapshot testing
*
*/
/*
 * Utilities
 */
/*
* Icon component
*
*/
/*
 * Utilities
 */
.dnb-icon {
  display: inline-block;
  vertical-align: middle;
  font-size: 1rem;
  line-height: 1rem;
  color: inherit;
  width: 1em;
  height: 1em;
}
.dnb-icon img, .dnb-core-style .dnb-icon img,
.dnb-icon svg,
.dnb-core-style .dnb-icon svg {
  width: inherit;
  height: inherit;
  shape-rendering: geometricprecision;
  vertical-align: top;
}
.dnb-icon svg[width="100%"] {
  width: inherit;
}
.dnb-icon svg[height="100%"] {
  height: inherit;
}
.dnb-icon--inherit-color svg:not([fill]), .dnb-icon--inherit-color svg [fill] {
  fill: currentcolor;
}
.dnb-icon--inherit-color svg [stroke] {
  stroke: currentcolor;
}
.dnb-icon--small {
  font-size: 0.75rem;
}
.dnb-icon--default {
  font-size: 1rem;
}
.dnb-icon--medium {
  font-size: 1.5rem;
}
.dnb-icon--large {
  font-size: 2rem;
}
.dnb-icon--x-large {
  font-size: 2.5rem;
}
.dnb-icon--xx-large {
  font-size: 3rem;
}
.dnb-icon--custom-size {
  width: auto;
  height: auto;
}
.dnb-icon--auto {
  font-size: 1em;
}
.dnb-icon--auto > .dnb-icon--wrapper {
  display: inline-flex;
  align-items: center;
  justify-content: center;
}
h1 > .dnb-icon, h2 > .dnb-icon, h3 > .dnb-icon, h4 > .dnb-icon, h5 > .dnb-icon, h6 > .dnb-icon {
  vertical-align: middle;
}
p > .dnb-icon {
  vertical-align: inherit;
}
.dnb-icon.dnb-skeleton {
  color: var(--skeleton-color) !important;
}
.dnb-icon.dnb-skeleton::before, .dnb-icon.dnb-skeleton::after {
  content: none !important;
}

/*
* Accordion component
*
*/
/*
 * Utilities
 */
.dnb-accordion {
  --accordion-border-width: 0.0625rem;
  --accordion-border-radius: 0.25rem;
  --accordion-easing: var(--easing-default);
  --accordion-header-margin-vertical: 1.5rem;
  --accordion-header-margin-vertical--description: 1rem;
  --accordion-header-wrapper-margin: 2rem;
  --accordion-header-wrapper-margin--icon-right: 1rem;
  --accordion-header-icon-gutter: 1rem;
  --accordion-header-icon-margin: 1rem;
  position: relative;
  display: flex;
  flex-direction: column;
  font-size: var(--font-size-small);
  line-height: var(--line-height-basis);
}
.dnb-accordion__header {
  display: flex;
  align-items: center;
  cursor: pointer;
  outline: none;
  border: var(--accordion-border-width) solid transparent;
  border-radius: var(--accordion-border-radius);
}
.dnb-accordion__header:focus {
  user-select: none;
}
.dnb-accordion__header * {
  pointer-events: none;
}
.dnb-accordion__header--prevent-click {
  user-select: none;
  pointer-events: none;
}
.dnb-accordion__header__wrapper {
  display: flex;
  flex-direction: column;
  width: 100%;
  margin: var(--accordion-header-margin-vertical) var(--accordion-header-wrapper-margin) var(--accordion-header-margin-vertical) 0;
}
.dnb-accordion__header--description .dnb-accordion__header__wrapper {
  margin-top: var(--accordion-header-margin-vertical--description);
  margin-bottom: var(--accordion-header-margin-vertical--description);
}
.dnb-accordion__header__container {
  margin: 0 1rem;
}
.dnb-accordion__header__icon {
  align-self: flex-start;
  margin: var(--accordion-header-margin-vertical) var(--accordion-header-icon-gutter) var(--accordion-header-margin-vertical) var(--accordion-header-icon-margin);
  font-size: var(--font-size-small);
  line-height: var(--line-height-small);
  transition: transform 400ms var(--accordion-easing);
}
.dnb-accordion__header--description .dnb-accordion__header__icon {
  margin-top: 1rem;
}
.dnb-accordion__header--icon-right {
  justify-content: space-between;
}
.dnb-accordion__header--icon-right .dnb-accordion__header__icon {
  margin-right: var(--accordion-header-icon-margin);
  margin-left: var(--accordion-header-icon-gutter);
  order: 3;
}
.dnb-accordion__header--icon-right .dnb-accordion__header__container {
  order: 1;
}
.dnb-accordion__header--icon-right .dnb-accordion__header__wrapper {
  order: 2;
  margin-right: 0;
  margin-left: var(--accordion-header-wrapper-margin--icon-right);
}
.dnb-accordion__header--icon-right .dnb-accordion__header__wrapper + .dnb-accordion__header__container {
  margin-right: 0;
}
.dnb-accordion__header--icon-right .dnb-accordion__header__container + .dnb-accordion__header__wrapper {
  margin-left: 0;
}
.dnb-accordion__header__title {
  font-size: var(--font-size-basis);
  line-height: var(--line-height-basis);
}
.dnb-accordion__header__description {
  font-size: var(--font-size-small);
  line-height: var(--line-height-small);
  font-weight: var(--font-weight-basis);
}
.dnb-accordion__header__title + .dnb-accordion__header__description {
  margin-top: 0.25rem;
}
.dnb-accordion__header__description + .dnb-accordion__header__title {
  margin-top: 0.25rem;
}
.dnb-accordion--expanded > .dnb-accordion__header .dnb-accordion__header__icon {
  transform: rotate(-180deg);
}
.dnb-accordion__content {
  display: flex;
  width: 100%;
  will-change: height;
  transition: height 400ms var(--accordion-easing), opacity 600ms var(--accordion-easing);
}
.dnb-accordion__content__inner {
  width: 100%;
  margin-top: 1rem;
}
.dnb-accordion__content.dnb-height-animation--hidden .dnb-accordion__content__inner {
  display: none;
}
.dnb-accordion-group--single-container {
  transition: min-height 1s var(--accordion-easing);
}
.dnb-accordion-group--single-container .dnb-accordion {
  position: static;
  max-width: 60rem;
}
@media screen and (min-width: 40em) {
  .dnb-accordion-group--single-container .dnb-accordion__header {
    width: 40%;
  }
  .dnb-accordion-group--single-container .dnb-accordion__content {
    margin-left: 1rem;
    transition: opacity 1s var(--accordion-easing);
  }
  .dnb-accordion-group--single-container .dnb-accordion__content__inner {
    margin-top: 0;
  }
}
@media screen and (min-width: 40em) {
  .dnb-accordion-group--single-container .dnb-accordion > .dnb-accordion__header .dnb-accordion__header__icon {
    transform: rotate(-90deg);
  }
}
.dnb-accordion-group--single-container .dnb-accordion-group__children {
  max-width: 60rem;
}
@media screen and (min-width: 40em) {
  .dnb-accordion-group--single-container .dnb-accordion-group__children {
    position: relative;
    display: flex;
    flex-direction: column;
  }
  .dnb-accordion-group--single-container .dnb-accordion-group__children .dnb-accordion__content {
    position: absolute;
    z-index: 10;
    top: 0;
    right: 0;
    width: 60%;
  }
}
.dnb-accordion > .dnb-accordion__header--no-animation .dnb-accordion__header__icon, html[data-visual-test] .dnb-accordion .dnb-accordion__header .dnb-accordion__header__icon {
  transition: none;
}"
`;

exports[`Accordion scss have to match default theme snapshot 1`] = `
"/*
* Accordion theme
*
*/
/*
 * Utilities
 */
/*
* Button mixins
*
*/
.dnb-accordion__header {
  border: none;
}
.dnb-accordion__header__icon, .dnb-accordion__header__container {
  color: var(--color-sea-green);
}
.dnb-accordion__header__description {
  color: var(--color-black-55);
}
.dnb-accordion--expanded > .dnb-accordion__header .dnb-accordion__header__title {
  font-weight: var(--font-weight-medium);
}
.dnb-accordion__variant--outlined > .dnb-accordion__header {
  color: var(--color-black-80);
  background-color: var(--color-white);
  --border-color: var(--color-black-8);
  --border-width: 0.0625rem;
  box-shadow: inset 0 0 0 var(--border-width) var(--border-color);
  border-color: transparent;
}
.dnb-accordion__variant--outlined > .dnb-accordion__header:focus[disabled], html:not([data-whatintent=touch]) .dnb-accordion__variant--outlined > .dnb-accordion__header:focus[disabled] {
  cursor: not-allowed;
}
.dnb-accordion__variant--outlined > .dnb-accordion__header:focus:not([disabled]), html:not([data-whatintent=touch]) .dnb-accordion__variant--outlined > .dnb-accordion__header:focus:not([disabled]) {
  outline: none;
}
html[data-whatinput=keyboard] .dnb-accordion__variant--outlined > .dnb-accordion__header:focus:not([disabled]), html[data-whatinput=keyboard] html:not([data-whatintent=touch]) .dnb-accordion__variant--outlined > .dnb-accordion__header:focus:not([disabled]) {
  color: var(--color-emerald-green);
  background-color: var(--color-white);
}
html[data-whatinput=keyboard] .dnb-accordion__variant--outlined > .dnb-accordion__header:focus:not([disabled]), html[data-whatinput=keyboard] html:not([data-whatintent=touch]) .dnb-accordion__variant--outlined > .dnb-accordion__header:focus:not([disabled]) {
  --border-color: var(--focus-ring-color);
  --border-width: var(--focus-ring-width);
  box-shadow: inset 0 0 0 var(--border-width) var(--border-color);
  border-color: transparent;
}
html:not([data-whatintent=touch]) .dnb-accordion__variant--outlined > .dnb-accordion__header:hover[disabled] {
  cursor: not-allowed;
}
html:not([data-whatintent=touch]) .dnb-accordion__variant--outlined > .dnb-accordion__header:hover:not([disabled]) {
  color: var(--color-emerald-green);
  background-color: var(--color-white);
  --border-color: var(--color-emerald-green);
  --border-width: 0.125rem;
  box-shadow: 0 0 0 var(--border-width) var(--border-color);
  border-color: transparent;
}
.dnb-accordion__variant--outlined > .dnb-accordion__header:active[disabled], html:not([data-whatintent=touch]) .dnb-accordion__variant--outlined > .dnb-accordion__header:active[disabled] {
  cursor: not-allowed;
}
.dnb-accordion__variant--outlined > .dnb-accordion__header:active:not([disabled]), html:not([data-whatintent=touch]) .dnb-accordion__variant--outlined > .dnb-accordion__header:active:not([disabled]) {
  color: var(--color-emerald-green);
  background-color: var(--color-pistachio);
  --border-color: var(--color-emerald-green);
  --border-width: 0.0625rem;
  box-shadow: 0 0 0 var(--border-width) var(--border-color);
  border-color: transparent;
}
.dnb-accordion__variant--outlined > .dnb-accordion__header[disabled] {
  background-color: var(--color-white);
  --border-color: var(--color-sea-green-30);
  --border-width: 0.0625rem;
  box-shadow: 0 0 0 var(--border-width) var(--border-color);
  border-color: transparent;
}
.dnb-accordion__variant--outlined > .dnb-accordion__header[disabled] * {
  color: var(--color-sea-green-30);
}
.dnb-accordion__variant--outlined.dnb-accordion--expanded > .dnb-accordion__header {
  color: var(--color-white);
  background-color: var(--color-sea-green);
  --border-color: transparent;
  --border-width: 0.0625rem;
  box-shadow: 0 0 0 var(--border-width) var(--border-color);
  border-color: transparent;
}
.dnb-accordion__variant--outlined.dnb-accordion--expanded > .dnb-accordion__header[disabled] {
  background-color: var(--color-sea-green-30);
}
.dnb-accordion__variant--outlined.dnb-accordion--expanded > .dnb-accordion__header[disabled] * {
  color: var(--color-white);
}
html[data-whatinput=keyboard] .dnb-accordion__variant--outlined.dnb-accordion--expanded > .dnb-accordion__header:not([disabled]):not(:active):not(:hover):focus {
  background-color: var(--color-mint-green);
}
html[data-whatinput=keyboard] .dnb-accordion__variant--outlined.dnb-accordion--expanded > .dnb-accordion__header:not([disabled]):not(:active):not(:hover):focus * {
  color: var(--color-emerald-green);
}
.dnb-accordion__variant--outlined.dnb-accordion--expanded > .dnb-accordion__header:not([disabled]):not(:active):not(:hover):focus {
  outline: none;
}
html[data-whatinput=keyboard] .dnb-accordion__variant--outlined.dnb-accordion--expanded > .dnb-accordion__header:not([disabled]):not(:active):not(:hover):focus {
  --border-color: var(--color-emerald-green);
  --border-width: var(--focus-ring-width);
  box-shadow: inset 0 0 0 var(--border-width) var(--border-color);
  border-color: transparent;
}
.dnb-accordion__variant--outlined.dnb-accordion--expanded > .dnb-accordion__header:not(:hover):not([disabled]) * {
  color: var(--color-white);
}
html:not([data-whatintent=touch]) .dnb-accordion__variant--outlined.dnb-accordion--expanded > .dnb-accordion__header.dnb-accordion--hover:not(:active):hover[disabled] {
  cursor: not-allowed;
}
html:not([data-whatintent=touch]) .dnb-accordion__variant--outlined.dnb-accordion--expanded > .dnb-accordion__header.dnb-accordion--hover:not(:active):hover:not([disabled]) {
  color: var(--color-white);
  background-color: var(--color-sea-green);
  --border-color: var(--color-sea-green);
  --border-width: 0.0625rem;
  box-shadow: 0 0 0 var(--border-width) var(--border-color);
  border-color: transparent;
}
.dnb-accordion__variant--outlined.dnb-accordion--expanded > .dnb-accordion__header.dnb-accordion--hover:not(:active) * {
  color: var(--color-white);
}"
`;
